<a href="https://colab.research.google.com/github/tada20001/NLP_2023/blob/main/CH19_02_%EC%9E%A0%EC%9E%AC_%EB%94%94%EB%A6%AC%ED%81%B4%EB%A0%88_%ED%95%A0%EB%8B%B9(Latent_Dirichlet_Allocation%2C_LDA).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

토픽 모델링은 문서의 집합에서 토픽을 찾아내는 프로세스를 말함. LDA는 대표적인 알고리즘임.

문서들은 토픽들의 혼합으로 구성되어 있는데, 각 토픽은 확률분포에 기반하여 단어들을 생성한다고 가정함. 데이터가 주어지면 LDA는 문서가 생성되었던 과정을 역추적함.


* 참고 링크 : https://lettier.com/projects/lda-topic-modeling/


### 1. 잠재 디리클레 할당(Latent Dirichlet Allocation, LDA) 개요
----------------
우선 LDA에 문서 집합을 입력하면 어떤 결과를 보여주는지 간소화된 예를 살펴보자. 아래와 같이 3개 문서에서 토픽을 찾는다고 하자.

* 문서1 : 저는 사과랑 바나나를 먹어요
* 문서2 : 우리는 귀여운 강아지가 좋아요
* 문서3 : 저의 깜찍하고 귀여운 강아지가 바나나를 먹어요

LDA를 수행할 때 문서 집합에서 토픽이 몇개가 존재할지 가정하는 것은 사용자가 직접 해야 함. 여기서는 LDA에 2개의 토픽을 찾는것으로 함. 즉 k=2. 

LDA가 위의 세문서로부터 2개의 토픽을 찾은 결과는 다음과 같음. 여기서는 LDA 입력 전에 전처리 처리를 했다고 가정함. 


LDA는 각 문서의 토픽분포와 각 토픽 내의 단어 분포를 추정함

<각 문서의 토픽 분포>

* 문서1 : 토픽 A 100%
* 문서2 : 토픽 B 100%
* 문서3 : 토픽 B 60%, 토픽 A 40%


<각 토픽의 단어 분포>
* 토픽A : 사과 20%, 바나나 40%, 먹어요 40%, 귀여운 0%, 강아지 0%, 깜찍하고 0%, 좋아요 0%
* 토픽B : 사과 0%, 바나나 0%, 먹어요 0%, 귀여운 33%, 강아지 33%, 깜찍하고 16%, 좋아요 16%


LDA는 토픽의 제목을 정해주지 않지만, 이 시점에서 알고리즘의 사용자는 위 결과로부터 두 토픽이 각각 과일에 대한 토픽과 강아지에 대한 토픽이라고 판단할 수 있음. 이제 LDA에 대해 알아봄.


### 2. LDA의 가정
----------------
LDA는 문서의 집합으로부터 어떤 토픽이 존재하는지 알아내기 위한 알고리즘임. LDA는 빈도수 기반의 표현방법인 BoW 행렬 DTM 또는 TF-IDF 행렬을 입력으로 하는데, 다만 문제는 LDA는 단어의 순서를 고려하지 않는다는 것임.

모든 문서 하나 하나가 작성될 때 다음과 같은 과정을 거쳐 작성되었다고 가정함

1) 문서에 사용할 단어의 개수 N을 정함 

2) 문서에 사용할 토픽의 혼합을 확률분포에 기반하여 결정함

3) 문서에 사용할 각 단어를 아래와 같이 정함
  * 토픽 분포에서 토픽 T를 확률적으로 고름
  * 선택한 토픽 T에서 단어의 출현 확률 분포에 기반해 문서에 사용할 단어를 고름


이러한 과정을 통해 문서가 작성되었다는 가정 하에 LDA는 토픽을 뽑아내기 위해 위 과정을 역으로 추적하는 역공학(reverse engineering)을 수행함

### 3. LDA 수행하기
-----------------
LDA 프로세스 정리.

1) 사용자는 알고리즘에게 토픽의 개수 k를 알려줌 : 토픽의 개수가 정해지면, k개의 토픽이 M개의 전체 문서에 걸쳐 분포되어 있다고 가정함

2) 모든 단어를 k개 토픽 중 하나에 각각 할당함 : 랜덤으로 할당하기 때문에 한단어가 한문서에 2회 이상 등장하면 각 단어는 서로 다른 토픽에 할당될 수 있음

3) 모든 문서의 모든 단어에 아래의 사항을 반복함

* 임의의 문서의 각 단어 w는 자신은 잘못된 토픽에 할당되어 있지만, 다른 단어들은 올바른 토픽에 할당되어져 있다고 가정함. 이에, 단어 w는 아래의 두가지 기준에 따라 토픽이 재할당됨
  * p(topic t | document d) : 문서 d의 단어들 중 토픽 t에 해당하는 단어들의 비
  * p(word w | topic t) : 각 토픽들 t에서 해당 단어 w의 분포


### 4. 잠재 디리클레 할당(LDA)과 잠재의미 분석(LSA)의 차이
--------------------------------------
* LSA : DTM을 차원 축소하여 축소차원에서 근접 단어들을 토픽으로 묶음
* LDA : 단어가 특정 토픽에 존재할 확률과 문서에 특정 토픽이 존재할 확률을 결합확률로 추정하여 토픽을 추출함


### 5. 실습을 통한 이해
---------------
LDA는 gensim을 사용함


#### 1) 정수 인코딩과 단어 집합 만들기


In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import fetch_20newsgroups
import nltk
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD

In [2]:
datasets = fetch_20newsgroups(shuffle=True, random_state=1, remove=('headers', 'footers', 'quotes'))
documents = datasets.data
len(documents)

11314

In [3]:
# 전처리
news_df = pd.DataFrame({'document': documents})
news_df['clean_doc'] = news_df['document'].str.replace('[^a-zA-Z]', ' ')

# 길이가 3이하 제거
news_df['clean_doc'] = news_df['clean_doc'].apply(lambda x: ' '.join([w for w in x.split() if len(w) > 3]))

# 전체 단어에 소문자 변환
news_df['clean_doc'] = news_df['clean_doc'].apply(lambda x: x.lower())


<ipython-input-3-b496bddd358c>:3: FutureWarning: The default value of regex will change from True to False in a future version.
  news_df['clean_doc'] = news_df['document'].str.replace('[^a-zA-Z]', ' ')


In [4]:
# 확인
news_df['clean_doc'][1]

'yeah expect people read actually accept hard atheism need little leap faith jimmy your logic runs steam sorry pity sorry that have these feelings denial about faith need well just pretend that will happily ever after anyway maybe start newsgroup atheist hard bummin much forget your flintstone chewables bake timmons'

In [5]:
tokenized_doc = news_df['clean_doc'].apply(lambda x: x.split())  # 토큰화

In [19]:
# 불용어 제거
nltk.download('stopwords')
stop_words = stopwords.words('english')
tokenized_doc = tokenized_doc.apply(lambda x: [item for item in x if item not in stop_words])

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [20]:
# 불용어 제거

tokenized_doc[:5]

0    [well, sure, story, seem, biased, disagree, st...
1    [yeah, expect, people, read, actually, accept,...
2    [although, realize, principle, strongest, poin...
3    [notwithstanding, legitimate, fuss, proposal, ...
4    [well, change, scoring, playoff, pool, unfortu...
Name: clean_doc, dtype: object

In [21]:
# 정수인코딩과 각 뉴스의 단어 빈도수 도출(word_id, word_frequency)
from gensim import corpora

In [22]:
dictionary = corpora.Dictionary(tokenized_doc)
corpus = [dictionary.doc2bow(text) for text in tokenized_doc]
print(corpus[1])  #(정수인코딩, 빈도수) 튜플로 도출

[(52, 1), (55, 1), (56, 1), (57, 1), (58, 1), (59, 1), (60, 1), (61, 1), (62, 1), (63, 1), (64, 1), (65, 1), (66, 2), (67, 1), (68, 1), (69, 1), (70, 1), (71, 2), (72, 1), (73, 1), (74, 1), (75, 1), (76, 1), (77, 1), (78, 2), (79, 1), (80, 1), (81, 1), (82, 1), (83, 1), (84, 1), (85, 2), (86, 1), (87, 1), (88, 1), (89, 1)]


In [23]:
# 단어별 정수값 확인 방법
print(dictionary[60])

bake


In [24]:
print(tokenized_doc[1])

['yeah', 'expect', 'people', 'read', 'actually', 'accept', 'hard', 'atheism', 'need', 'little', 'leap', 'faith', 'jimmy', 'logic', 'runs', 'steam', 'sorry', 'pity', 'sorry', 'feelings', 'denial', 'faith', 'need', 'well', 'pretend', 'happily', 'ever', 'anyway', 'maybe', 'start', 'newsgroup', 'atheist', 'hard', 'bummin', 'much', 'forget', 'flintstone', 'chewables', 'bake', 'timmons']


#### 2) LDA 모델 훈련

기존의 뉴스 데이터가 총 20개의 카테고리를 가지고 있었으므로 토픽의 개수를 20으로 하여 LDA 모델을 학습함

In [25]:
import gensim
NUM_TOPICS = 20
ldamodel = gensim.models.ldamodel.LdaModel(corpus, num_topics=NUM_TOPICS, id2word=dictionary, passes=15)
topics = ldamodel.print_topics(num_words=4)
for topic in topics:
  print(topic)

(0, '0.023*"please" + 0.023*"mail" + 0.018*"send" + 0.017*"email"')
(1, '0.022*"game" + 0.022*"team" + 0.016*"games" + 0.014*"play"')
(2, '0.009*"bike" + 0.007*"cars" + 0.007*"engine" + 0.005*"road"')
(3, '0.020*"encryption" + 0.017*"chip" + 0.016*"keys" + 0.015*"clipper"')
(4, '0.013*"drive" + 0.011*"system" + 0.009*"card" + 0.008*"disk"')
(5, '0.013*"nist" + 0.012*"crypt" + 0.010*"ncsl" + 0.010*"colorado"')
(6, '0.011*"sabbath" + 0.011*"rockefeller" + 0.009*"allocation" + 0.009*"mask"')
(7, '0.041*"space" + 0.016*"nasa" + 0.009*"data" + 0.009*"launch"')
(8, '0.023*"ground" + 0.013*"wiring" + 0.013*"neutral" + 0.012*"circuit"')
(9, '0.006*"pain" + 0.006*"water" + 0.006*"food" + 0.006*"years"')
(10, '0.011*"people" + 0.010*"would" + 0.005*"government" + 0.005*"many"')
(11, '0.012*"people" + 0.010*"would" + 0.009*"jesus" + 0.008*"believe"')
(12, '0.013*"president" + 0.011*"university" + 0.009*"national" + 0.008*"april"')
(13, '0.015*"armenian" + 0.013*"israel" + 0.013*"jews" + 0.010*"tu

각 단어 앞에 붙은 수치는 단어의 해당 토픽에 대한 기여도를 보여줌. 또한, 맨 앞에 있는 토픽번호는 0부터 시작하므로 총 20개의 토픽은 0부터 19까지 번호가 할당되어져 있음. 


passes는 알고리즘의 동작횟수를 말하는데, 알고리즘이 결정하는 토픽의 값이 적절히 수렴할 수 있도록 충분히 적당한 횟수를 정해주면 됨.

여기에서는 총 15회를 수행하였음. 그리고 num_words=4로 4개의 단어만 출력하도록 함.

만약 10개 단어를 출력하고 싶으면 아래의 토드를 수행하면 됨

In [26]:
topics = ldamodel.print_topics(num_words=10)
for topic in topics:
  print(topic)

(0, '0.023*"please" + 0.023*"mail" + 0.018*"send" + 0.017*"email" + 0.015*"thanks" + 0.014*"list" + 0.014*"information" + 0.011*"address" + 0.011*"info" + 0.008*"anyone"')
(1, '0.022*"game" + 0.022*"team" + 0.016*"games" + 0.014*"play" + 0.013*"season" + 0.011*"league" + 0.011*"hockey" + 0.011*"players" + 0.009*"period" + 0.008*"teams"')
(2, '0.009*"bike" + 0.007*"cars" + 0.007*"engine" + 0.005*"road" + 0.005*"cover" + 0.005*"miles" + 0.005*"front" + 0.005*"wire" + 0.005*"insurance" + 0.004*"right"')
(3, '0.020*"encryption" + 0.017*"chip" + 0.016*"keys" + 0.015*"clipper" + 0.013*"security" + 0.012*"privacy" + 0.011*"government" + 0.010*"public" + 0.009*"system" + 0.009*"technology"')
(4, '0.013*"drive" + 0.011*"system" + 0.009*"card" + 0.008*"disk" + 0.007*"problem" + 0.007*"used" + 0.006*"know" + 0.006*"like" + 0.006*"need" + 0.006*"also"')
(5, '0.013*"nist" + 0.012*"crypt" + 0.010*"ncsl" + 0.010*"colorado" + 0.009*"plaintext" + 0.007*"zone" + 0.007*"marc" + 0.007*"patent" + 0.006*"jo

#### 3) LDA 시각화하기
시각화를 위해 pyLDAvis 설치가 필요함. 

In [39]:
!pip install pyLDAvis

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [40]:
import pyLDAvis
import pyLDAvis.gensim_models

In [41]:
pyLDAvis.enable_notebook()
vis = pyLDAvis.gensim_models.prepare(ldamodel, corpus, dictionary)

#pyLDAvis.display(vis)

ERROR:concurrent.futures:exception calling callback for <Future at 0x7fd0cf2cf040 state=finished raised BrokenProcessPool>
joblib.externals.loky.process_executor._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py", line 391, in _process_worker
    call_item = call_queue.get(block=True, timeout=timeout)
  File "/usr/lib/python3.10/multiprocessing/queues.py", line 122, in get
    return _ForkingPickler.loads(res)
ModuleNotFoundError: No module named 'pandas.core.indexes.numeric'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/_base.py", line 26, in _invoke_callbacks
    callback(self)
  File "/usr/local/lib/python3.10/dist-packages/joblib/parallel.py", line 385, in __call__
    self.parallel.dispatch_next()
  File "/usr/local/lib/python3.10/dist-packages/joblib/p

BrokenProcessPool: ignored

#### 4) 문서별 토픽 분포 보기

각 문서의 토픽 분포는 이미 훈련된 LDA 모델인 ldamodel[]에 전체 데이터가 정수 인코딩된 결과를 넣은 후에 확인이 가능함. 여기에서는 상위 5개 문서만 확인함.

In [42]:
for i, topic_list in enumerate(ldamodel[corpus]):
  if i == 5:
    break
  print(i, '번째 문서의 topic 비율은', topic_list)

0 번째 문서의 topic 비율은 [(4, 0.023191622), (5, 0.049878158), (10, 0.14050865), (11, 0.33026433), (13, 0.41948438), (18, 0.025173567)]
1 번째 문서의 topic 비율은 [(9, 0.05978697), (11, 0.49811935), (14, 0.030098785), (15, 0.02761466), (18, 0.36461803)]
2 번째 문서의 topic 비율은 [(1, 0.046760786), (5, 0.017713374), (10, 0.16278163), (13, 0.29913965), (18, 0.46087527)]
3 번째 문서의 topic 비율은 [(2, 0.05596703), (3, 0.32127374), (4, 0.063700855), (10, 0.1889689), (18, 0.27894858), (19, 0.0800175)]
4 번째 문서의 topic 비율은 [(0, 0.111899436), (1, 0.34376517), (11, 0.36742392), (18, 0.1472693)]


In [44]:
dictionary[4]

'biased'

위의 출력 결과에서 (숫자, 확률)은 각각 토픽번호와 해당 토픽이 해당 문서에서 차지하는 분포도를 의미함. 예를 들어, 0번째 문서의 토픽 비율에서 (4, 0.023191622)는 4번 토픽이 23% 분포를 가지는 것을 의미함. 이를 데이터 프레임형식으로 출력함.

In [75]:
def make_topictable_per_doc(ldamodel, corpus):
  topic_list_new = []

  for i, topic_list in enumerate(ldamodel[corpus]):
    doc = topic_list[0] if ldamodel.per_word_topics else topic_list
    doc = sorted(doc, key=lambda x: (x[1]), reverse=True)  # 비중 내림차순
    
    for j, (topic_num, prop_topic) in enumerate(doc):
      if j == 0:
        topic_list_new.append([int(topic_num), dictionary[int(topic_num)], round(prop_topic, 4), topic_list])
      else:
        break
  return pd.DataFrame(topic_list_new)

In [77]:
topictable = make_topictable_per_doc(ldamodel, corpus)
topictable = topictable.reset_index()
topictable.columns = ['문서번호', '가장 비중이 높은 토픽번호', '가장 비중이 높은 토픽', '가장 높은 토픽의 비중', '각 토픽의 비중']
topictable[:10]

문서번호  가장 비중이 높은 토픽번호 가장 비중이 높은 토픽  가장 높은 토픽의 비중  \
0     0              13    europeans        0.4197   
1     1              11     disagree        0.4980   
2     2              18       ignore        0.4609   
3     3               3         away        0.3213   
4     4              11     disagree        0.3674   
5     5              11     disagree        0.7139   
6     6               4       biased        0.3470   
7     7              18       ignore        0.3098   
8     8              11     disagree        0.2458   
9     9              18       ignore        0.4131   

                                            각 토픽의 비중  
0  [(4, 0.023135796), (5, 0.04987826), (10, 0.139...  
1  [(9, 0.059786715), (11, 0.49801734), (14, 0.03...  
2  [(1, 0.046760384), (5, 0.017713375), (10, 0.16...  
3  [(2, 0.05599415), (3, 0.32127926), (4, 0.06369...  
4  [(0, 0.11190306), (1, 0.34376398), (11, 0.3673...  
5               [(11, 0.71386826), (19, 0.24520636)]  
6  [(4, 0.34704548), (8, 0.11050618), (10, 0.1016...  
7  [(2, 0.18673246), (5, 0.0473757), (9, 0.021589...  
8  [(4, 0.21488655), (7, 0.03319339), (9, 0.13529...  
9  [(2, 0.2805147), (3, 0.09458226), (4, 0.157151...